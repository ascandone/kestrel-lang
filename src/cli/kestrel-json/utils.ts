import { KestrelJson } from "./schema";
import { existsSync } from "node:fs";
import { exit } from "node:process";
import { mkdir, readFile, writeFile } from "fs/promises";
import { join } from "path";
import { kestrelJsonDecoder } from "./decoder";
import * as dec from "ts-decode";

export const CONFIG_NAME = "kestrel.json";

export async function readConfig(
  root: string = process.cwd(),
): Promise<KestrelJson> {
  try {
    const f = await readFile(join(root, CONFIG_NAME));
    const res = kestrelJsonDecoder.decode(JSON.parse(f.toString()));
    if (res.error) {
      console.error(`Invalid config:\n${dec.reasonToXmlString(res.reason)}`);
      exit(1);
    } else {
      return res.value;
    }
  } catch {
    console.error(`Config not found`);
    exit(1);
  }
}

export async function writeConfig(path: string, config: unknown) {
  const kestrelJsonPath = join(path, CONFIG_NAME);
  const configExists = existsSync(kestrelJsonPath);
  if (configExists) {
    console.error(`Config already exists`);
    exit(1);
  }

  const content = JSON.stringify(config, null, 2);
  await writeFile(kestrelJsonPath, content + "\n");
  await mkdir(join(path, "src"));
}
