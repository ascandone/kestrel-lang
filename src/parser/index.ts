export { parse, unsafeParse, ParseResult } from "./parser";
