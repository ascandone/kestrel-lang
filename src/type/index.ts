export * from "./type";
export { typePPrint } from "./prettyPrint";
