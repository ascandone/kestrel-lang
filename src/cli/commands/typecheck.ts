import { readFileSync } from "fs";
import { parse } from "../../parser";
import { typecheck } from "../../typecheck/typecheck";
import { typeErrorPPrint } from "../../typecheck/pretty-printer";
import { prelude } from "../../typecheck/prelude";
import { Span } from "../../ast";

const FgRed = "\x1b[31m";
const Reset = "\x1b[0m";
const FgBlack = "\x1b[30m";
const BgWhite = "\x1b[47m";

export function typecheckCmd(path: string) {
  const f = readFileSync(path);

  const src = f.toString();
  const parseResult = parse(src);
  if (!parseResult.ok) {
    console.log(
      `${FgRed}Parsing error:${Reset} ${parseResult.matchResult.message!}`,
    );
    return;
  }

  const [, errors] = typecheck(parseResult.value, prelude);

  for (const error of errors) {
    const msg = typeErrorPPrint(error);
    console.log(`${FgRed}Error:${Reset} ${msg}`);
    console.log(showErrorLine(src, error.node.span), "\n\n");
  }

  if (errors.length > 0) {
    const plErr = errors.length === 1 ? "error" : "errors";
    console.log(`[Found ${errors.length} ${plErr}]\n`);
  } else {
    console.log("Found no errors âœ…");
  }
}

export function showErrorLine(src: string, [start, end]: Span): string {
  const startPos = offsetToPosition(src, start);
  const endPos = offsetToPosition(src, end);
  const lines = src.split("\n");

  // TODO handle nums with many digits
  const digitsPadding = repeatN(" ", 1);

  function showLine(line: number) {
    const codeLine = lines[line]!;
    const lineNum = `${BgWhite}${FgBlack}${line + 1}${Reset}`;
    return `${lineNum} ${codeLine}`;
  }

  function showErr([start, end]: [start: number, end: number]) {
    const errPadding = repeatN(" ", start);
    const errHighlight = repeatN("~", end - start);
    return `${BgWhite}${digitsPadding}${Reset} ${FgRed}${errPadding}${errHighlight}${Reset}`;
  }

  const ret: string[] = [];
  for (
    let currentLine = startPos.line;
    currentLine <= endPos.line;
    currentLine++
  ) {
    const startChar = currentLine === startPos.line ? startPos.character : 0;

    const endChar =
      currentLine === endPos.line
        ? endPos.character
        : lines[currentLine]!.length;

    ret.push(showLine(currentLine), showErr([startChar, endChar]));
  }

  return ret.join("\n");
}

type Position = { line: number; character: number };

function offsetToPosition(src: string, offset: number): Position {
  let line = 0,
    character = 0;

  for (const ch of src) {
    if (offset === 0) {
      break;
    }

    offset--;
    if (ch === "\n") {
      line++;
      character = 0;
    } else {
      character++;
    }
  }

  return { line, character };
}

function repeatN(ch: string, times: number) {
  return Array.from({ length: times }, () => ch).join("");
}
