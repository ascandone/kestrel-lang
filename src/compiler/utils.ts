import * as t from "@babel/types";
import { TypedTypeDeclaration } from "../typecheck";

export const TAG_FIELD: t.Identifier = { type: "Identifier", name: "$" };

export function sanitizeNamespace(ns: string): string {
  return ns?.replace(/\//g, "$");
}

export function joinAndExprs(exprs: t.Expression[]): t.Expression {
  if (exprs.length === 0) {
    return { type: "BooleanLiteral", value: true };
  }

  return exprs.reduce(
    (left, right): t.Expression => ({
      type: "LogicalExpression",
      operator: "&&",
      left,
      right,
    }),
  );
}

// TODO utils file was a bad idea
export type AdtReprType = "default" | "enum" | "unboxed";
export function getAdtReprType(
  decl: TypedTypeDeclaration & { type: "adt" },
): AdtReprType {
  const isEnum = decl.variants.every((v) => v.args.length === 0);
  if (isEnum) {
    return "enum";
  }

  return "default";
}
