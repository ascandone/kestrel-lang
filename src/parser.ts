/* eslint-disable @typescript-eslint/no-unused-vars */
import grammar from "./parser/grammar.ohm-bundle";
import {
  ConstLiteral,
  Program,
  Expr,
  Declaration,
  Span,
  SpanMeta,
  TypeAst,
  TypeDeclaration,
  TypeVariant,
  MatchPattern,
} from "./ast";
import type {
  IterationNode,
  MatchResult,
  NonterminalNode,
  Node as OhmNode,
  TerminalNode,
} from "ohm-js";
import { TypeMeta } from "./typecheck/typecheck";

function getSpan({ source }: OhmNode): Span {
  return [source.startIdx, source.endIdx];
}

const semantics = grammar.createSemantics();

semantics.addOperation<ConstLiteral>("lit()", {
  number_whole(node) {
    return { type: "int", value: Number(node.sourceString) };
  },

  number_fract(_intPart, _comma, _floatPart) {
    return { type: "float", value: Number(this.sourceString) };
  },

  string(_lDel, _s, _rDel) {
    return { type: "string", value: JSON.parse(this.sourceString) };
  },
});

function parseInfix(
  this: NonterminalNode,
  left: NonterminalNode,
  op: TerminalNode,
  right: NonterminalNode,
): Expr<SpanMeta> {
  return {
    type: "application",
    span: getSpan(this),
    caller: {
      type: "identifier",
      name: op.sourceString,
      span: getSpan(op),
    },
    args: [left.expr(), right.expr()],
  };
}

function parsePrefix(
  this: NonterminalNode,
  op: TerminalNode,
  exp: NonterminalNode,
): Expr<SpanMeta> {
  return {
    type: "application",
    span: getSpan(this),
    caller: {
      type: "identifier",
      name: op.sourceString,
      span: getSpan(op),
    },
    args: [exp.expr()],
  };
}

semantics.addOperation<{ name: string } & SpanMeta>("ident()", {
  infixIdent(_lparens, ident, _rparens) {
    return {
      name: ident.sourceString,
      span: getSpan(this),
    };
  },
  ident(_hd, _tl) {
    return {
      name: this.sourceString,
      span: getSpan(this),
    };
  },
});

semantics.addOperation<MatchPattern>("matchPattern()", {
  MatchPattern_ident(ident) {
    return {
      type: "ident",
      ident: ident.sourceString,
      span: getSpan(this),
    };
  },
  MatchPattern_lit(lit) {
    return {
      type: "lit",
      literal: lit.lit(),
      span: getSpan(this),
    };
  },

  ConsPattern_cons(l, cons, r) {
    return {
      type: "constructor",
      name: cons.sourceString,
      args: [l.matchPattern(), r.matchPattern()],
      span: getSpan(this),
    };
  },

  MatchPattern_tuple(_lparens, x, _comma, xs, _rparens) {
    const xs_ = xs.asIteration().children.map((c) => c.matchPattern());

    const count = 1 + xs_.length;
    const name = `Tuple${count}`;
    return {
      type: "constructor",
      name,
      args: [x.matchPattern(), ...xs_],
      span: getSpan(this),
    };
  },
  MatchPattern_constructor(ident, _lparent, args, _rparens) {
    let args_: MatchPattern[] = [];
    if (args.numChildren > 0) {
      args_ = args
        .child(0)
        .asIteration()
        .children.map((a) => a.matchPattern());
    }
    return {
      type: "constructor",
      name: ident.sourceString,
      args: args_,
      span: getSpan(this),
    };
  },
});

semantics.addOperation<[MatchPattern, Expr<TypeMeta>]>("matchClause()", {
  MatchClause_clause(match, _arrow, expr) {
    return [match.matchPattern(), expr.expr()];
  },
});

semantics.addOperation<Expr<SpanMeta>>("expr()", {
  PipeExp_pipe(left, _pipe, caller, _lparens, args, _trailingComma, _rparens) {
    return {
      type: "application",
      caller: caller.expr(),
      args: [left.expr(), ...args.asIteration().children.map((c) => c.expr())],
      span: getSpan(this),
    };
  },
  PriExp_literal(n) {
    return {
      type: "constant",
      value: n.lit(),
      span: getSpan(this),
    };
  },

  ConsExpr_cons(hd, cons, tl) {
    return {
      type: "application",
      span: getSpan(this),
      caller: {
        type: "identifier",
        name: "Cons",
        span: getSpan(cons),
      },
      args: [hd.expr(), tl.expr()],
    };
  },

  PriExp_tuple(_lparens, x, _comma, xs, _rparens) {
    const xs_ = xs.asIteration().children.map((x) => x.expr());

    const count = 1 + xs_.length;

    return {
      type: "application",
      caller: {
        type: "identifier",
        name: `Tuple${count}`,
        span: getSpan(this),
      },
      args: [x.expr(), ...xs_],
      span: getSpan(this),
    };
  },
  PriExp_paren(_open, e, _close) {
    return e.expr();
  },
  PriExp_constructor(node) {
    return {
      type: "identifier",
      name: node.sourceString,
      span: getSpan(node),
    };
  },

  ident(_hd, _tl) {
    return {
      type: "identifier",
      ...this.ident(),
    };
  },
  CompExp_lt: parseInfix,
  CompExp_gt: parseInfix,
  CompExp_gte: parseInfix,
  CompExp_lte: parseInfix,
  EqExpr_eq: parseInfix,
  EqExpr_neq: parseInfix,
  OrExpr_or: parseInfix,
  AndExpr_and: parseInfix,
  AddExp_plus: parseInfix,
  AddExp_plusFloat: parseInfix,
  AddExp_minus: parseInfix,
  AddExp_minusFloat: parseInfix,
  MulExp_mult: parseInfix,
  MulExp_multFloat: parseInfix,
  MulExp_divide: parseInfix,
  MulExp_divideFloat: parseInfix,
  MulExp_rem: parseInfix,
  ExpExp_power: parseInfix,
  AddExp_concat: parseInfix,

  NotExp_not: parsePrefix,

  PriExp_apply(f, _lpar, args, _trailingComma, _rpar) {
    return {
      type: "application",
      caller: f.expr(),
      args: args.asIteration().children.map((c) => c.expr()),
      span: getSpan(this),
    };
  },

  PriExp_block(_lbracket, block, _rbracket) {
    return block.expr();
  },

  PriExp_fn(_fn, params, _lbracket, block, _rbracket) {
    return {
      type: "fn",
      params: params.asIteration().children.map((c) => c.ident()),
      body: block.expr(),
      span: getSpan(this),
    };
  },

  PriExp_if(_if, cond, _lb1, x, _rb1, _else, _lb2, y, _rb2) {
    return {
      type: "if",
      condition: cond.expr(),
      then: x.expr(),
      else: y.expr(),
      span: getSpan(this),
    };
  },

  PriExp_match(_match, expr, _lbracket, clauses, _comma, _rbracket) {
    return {
      type: "match",
      expr: expr.expr(),
      clauses: clauses.asIteration().children.map((c) => c.matchClause()),
      span: getSpan(this),
    };
  },

  BlockContent_exp(e) {
    return e.expr();
  },

  BlockContent_monadicLet(_let, bindFIdent, ident, _eq, value, _comma, body) {
    return {
      type: "application",
      caller: bindFIdent.expr(),
      args: [
        value.expr(),
        {
          type: "fn",
          params: [ident.ident()],
          body: body.expr(),
          span: getSpan(this),
        },
      ],
      span: getSpan(this),
    };
  },

  BlockContent_let(_let, ident, _eq, value, _comma, body) {
    return {
      type: "let",
      binding: ident.ident(),
      value: value.expr(),
      body: body.expr(),
      span: getSpan(this),
    };
  },
});

semantics.addOperation<TypeAst>("type()", {
  Type_any(_underscore) {
    return { type: "any", span: getSpan(this) };
  },

  Type_var(ident) {
    return { type: "var", ident: ident.sourceString, span: getSpan(this) };
  },

  Type_fn(_fn, _lparens, args, _rparens, _arrow, ret) {
    const args_ = args.asIteration().children.map((arg) => arg.type());
    return { type: "fn", args: args_, return: ret.type(), span: getSpan(this) };
  },

  Type_named(ident, _lbracket, args, _rbracket) {
    let args_: TypeAst[] = [];
    if (args.numChildren > 0) {
      args_ = args
        .child(0)
        .asIteration()
        .children.map((c) => c.type());
    }
    return {
      type: "named",
      args: args_,
      name: ident.sourceString,
      span: getSpan(this),
    };
  },
});

type Statement =
  | { type: "typeDeclaration"; decl: TypeDeclaration }
  | { type: "declaration"; decl: Declaration };

semantics.addOperation<TypeVariant>("typeVariant()", {
  TypeVariant(name, _lparens, args, _rparens) {
    let args_: TypeAst[] = [];
    if (args.numChildren > 0) {
      args_ = args
        .child(0)
        .asIteration()
        .children.map((c) => c.type());
    }

    return { name: name.sourceString, args: args_ };
  },
});

function parseParams(params: IterationNode) {
  let params_: Array<{ name: string } & SpanMeta> = [];
  if (params.numChildren > 0) {
    params_ = params
      .child(0)
      .asIteration()
      .children.map((c) => ({
        name: c.sourceString,
        span: getSpan(c),
      }));
  }
  return params_;
}
semantics.addOperation<Statement>("statement()", {
  TypeDeclaration_externType(_extern, _type, typeName, _lT, params, _rT) {
    return {
      type: "typeDeclaration",
      decl: {
        type: "extern",
        params: parseParams(params),
        name: typeName.sourceString,
        span: getSpan(this),
      },
    };
  },

  TypeDeclaration_typeDef(
    _type,
    typeName,
    _lT,
    params,
    _rT,
    _lbracket,
    variants,
    _trailingComma,
    _rbracket,
  ) {
    const variants_ = variants
      .asIteration()
      .children.map<TypeVariant>((n) => n.typeVariant());

    return {
      type: "typeDeclaration",
      decl: {
        type: "adt",
        params: parseParams(params),
        name: typeName.sourceString,
        variants: variants_,
        span: getSpan(this),
      },
    };
  },
  Declaration_externLetStmt(_extern, _let, ident, _colon, typeHint) {
    return {
      type: "declaration",
      decl: {
        extern: true,
        binding: ident.ident(),
        span: getSpan(this),
        typeHint: {
          ...typeHint.type(),
          span: getSpan(typeHint.child(0)),
        },
      },
    };
  },
  Declaration_letStmt(_let, ident, _colon, typeHint, _eq, exp) {
    const th =
      typeHint.numChildren === 0
        ? {}
        : {
            typeHint: {
              ...typeHint.child(0).type(),
              span: getSpan(typeHint.child(0)),
            },
          };

    return {
      type: "declaration",
      decl: {
        extern: false,
        binding: ident.ident(),
        value: exp.expr(),
        span: getSpan(this),
        ...th,
      },
    };
  },
});

semantics.addOperation<Program>("parse()", {
  MAIN(statements) {
    const statements_ = statements.children.map<Statement>((child) =>
      child.statement(),
    );

    return {
      typeDeclarations: statements_.flatMap((st) =>
        st.type === "typeDeclaration" ? [st.decl] : [],
      ),
      declarations: statements_.flatMap((st) =>
        st.type === "declaration" ? [st.decl] : [],
      ),
    };
  },
});

export type ParseResult<T> =
  | { ok: true; value: T }
  | { ok: false; matchResult: MatchResult };

export function parse(input: string): ParseResult<Program<SpanMeta>> {
  const matchResult = grammar.match(input);
  if (matchResult.failed()) {
    return { ok: false, matchResult };
  }

  return { ok: true, value: semantics(matchResult).parse() };
}

export function unsafeParse(input: string): Program<SpanMeta> {
  const res = parse(input);
  if (res.ok) {
    return res.value;
  }

  throw new Error(res.matchResult.message!);
}
