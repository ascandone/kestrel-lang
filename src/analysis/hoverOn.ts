import { Position, RangeMeta, contains } from "../parser";
import { typeToString } from "../type";
import { Finder } from "../typecheck/astLookup";
import {
  IdentifierResolution,
  TypedDeclaration,
  TypedModule,
  TypedTypeAst,
  TypedTypeDeclaration,
} from "../typecheck/typedAst";
import { firstBy, statementByOffset } from "./common";

export type HoveredInfo =
  | IdentifierResolution
  | { type: "type"; typeDecl: TypedTypeDeclaration; namespace: string }
  | { type: "field"; type_: string };

export type Hovered = RangeMeta & { hovered: HoveredInfo };

/**
 * TODO fix regression:
 * we need to make sure that here:
 * ```kestrel
 * let f: (a) -> a = fn a {
 *   let id = fn b { b };
 *   a
 * }
 * ```
 *
 * we don't show `(a) -> a` when hovering on id
 */
export function hoverOn(
  package_: string,
  namespace: string,
  module: TypedModule,
  position: Position,
): Hovered | undefined {
  const statement = statementByOffset(module, position);
  if (statement === undefined) {
    return undefined;
  }

  switch (statement.type) {
    case "declaration": {
      if (statement.declaration.typeHint !== undefined) {
        const res = hoverOnTypeAst(
          statement.declaration.typeHint.mono,
          position,
        );
        if (res !== undefined) {
          return res;
        }
      }

      const d = hoverOnDecl(
        package_,
        namespace,
        statement.declaration,
        position,
      );
      if (d !== undefined) {
        return d;
      }
      return undefined;
    }

    case "type-declaration":
      if (statement.typeDeclaration.type === "adt") {
        for (const variant of statement.typeDeclaration.variants) {
          if (!contains(variant, position)) {
            continue;
          }

          const res = firstBy(variant.args, (arg) =>
            hoverOnTypeAst(arg, position),
          );
          if (res !== undefined) {
            return res;
          }

          return {
            range: variant.range,
            hovered: {
              type: "constructor",
              variant,
              declaration: statement.typeDeclaration,
              package_,
              namespace,
            },
          };
        }
      }

      return {
        range: statement.typeDeclaration.range,
        hovered: {
          type: "type",
          namespace,
          typeDecl: statement.typeDeclaration,
        },
      };

    case "import":
      return undefined;
  }
}

export function hoverOnTypeAst(
  typeAst: TypedTypeAst,
  position: Position,
): Hovered | undefined {
  if (!contains(typeAst, position)) {
    return;
  }

  switch (typeAst.type) {
    case "var":
    case "any":
      return undefined;
    case "named": {
      if (typeAst.$resolution === undefined) {
        return undefined;
      }

      const res = firstBy(typeAst.args, (arg) => hoverOnTypeAst(arg, position));
      if (res !== undefined) {
        return res;
      }

      return {
        range: typeAst.range,
        hovered: {
          type: "type",
          typeDecl: typeAst.$resolution.declaration,
          namespace: typeAst.$resolution.namespace,
        },
      };
    }

    case "fn": {
      return (
        firstBy(typeAst.args, (arg) => hoverOnTypeAst(arg, position)) ??
        hoverOnTypeAst(typeAst.return, position)
      );
    }
  }
}

export function hoverToMarkdown(
  // scheme: TypeScheme,
  { hovered }: Hovered,
): string {
  switch (hovered.type) {
    case "field": {
      return `\`\`\`
${hovered.type_}
\`\`\`
field
`;
    }

    case "type":
      return `\`\`\`
type ${hovered.typeDecl.name}
\`\`\`

${hovered.typeDecl.docComment ?? ""}
      `;
    case "local-variable": {
      const tpp = typeToString(hovered.binding.$type.asType());
      return `\`\`\`
${hovered.binding.name}: ${tpp}
\`\`\`
local declaration
`;
    }

    case "global-variable": {
      const tpp = typeToString(hovered.declaration.binding.$type.asType());
      return `\`\`\`
${hovered.declaration.binding.name}: ${tpp}
\`\`\`
${hovered.declaration.docComment ?? ""}
`;
    }

    case "constructor": {
      const tpp = typeToString(hovered.variant.$type);
      return `\`\`\`
${hovered.variant.name}: ${tpp}
\`\`\`
type constructor
`;
    }
  }
}

function hoverOnDecl(
  package_: string,
  namespace: string,
  declaration: TypedDeclaration,
  position: Position,
): Hovered | undefined {
  if (contains(declaration.binding, position)) {
    return {
      range: declaration.binding.range,
      hovered: {
        type: "global-variable",
        package_,
        declaration,
        namespace,
      },
    };
  }

  if (!declaration.extern && contains(declaration.value, position)) {
    return hoveredFinder(position).visitExpr(declaration.value);
  }

  return undefined;
}

function hoveredFinder(position: Position) {
  return new Finder<Hovered>(position, {
    onMatchPattern(pattern, next) {
      switch (pattern.type) {
        case "identifier":
          return {
            range: pattern.range,
            hovered: { type: "local-variable", binding: pattern },
          };

        default:
          return next();
      }
    },

    onExpression(expr, next) {
      switch (expr.type) {
        case "field-access":
          if (
            !contains(expr.field, position) ||
            expr.$resolution === undefined
          ) {
            break;
          }

          return {
            hovered: {
              type: "field",
              type_: typeToString(expr.$type.asType()),
            },
            range: expr.field.range,
          };

        case "identifier":
          if (expr.$resolution === undefined) {
            break;
          }

          return {
            range: expr.range,
            hovered: expr.$resolution,
          };
      }

      return next();
    },
  });
}
