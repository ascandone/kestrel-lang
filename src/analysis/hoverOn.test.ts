import { test, expect } from "vitest";
import { Hovered, hoverOn, hoverToMarkdown } from "./hoverOn";
import { unsafeParse } from "../parser";
import { typecheck } from "../typecheck/typecheck";
import { rangeOf } from "./__test__/utils";
import { resolveType } from "../type";

test("hover a declaration's binding", () => {
  const src = `let x = 42`;
  const hoverable = parseHover(src, "x")!;
  expect(hoverable).toEqual<Hovered>({
    range: rangeOf(src, "x"),
    hovered: expect.objectContaining({
      type: "global-variable",
      declaration: expect.objectContaining({
        binding: expect.objectContaining({
          name: "x",
        }),
      }),
    }),
  });
});

test("hover a declaration's binding (2d occurrence)", () => {
  const src = `let x = 42\nlet y = 100`;
  const hoverable = parseHover(src, "y")!;
  expect(hoverable).toEqual<Hovered>({
    range: rangeOf(src, "y"),
    hovered: expect.objectContaining({
      type: "global-variable",
      declaration: expect.objectContaining({
        binding: expect.objectContaining({
          name: "y",
        }),
      }),
    }),
  });
});

test("hover fn param", () => {
  const src = `let f = fn x { 0 }`;
  const hoverable = parseHover(src, "x")!;
  expect(hoverable).toEqual<Hovered>({
    range: rangeOf(src, "x"),
    hovered: expect.objectContaining({
      type: "local-variable",
      binding: expect.objectContaining({
        name: "x",
      }),
    }),
  });
});

test("hover let expr binding", () => {
  const src = `let f = {
      let x = 42;
      0
    }`;
  const hoverable = parseHover(src, "x")!;
  expect(hoverable).toEqual<Hovered>({
    range: rangeOf(src, "x"),
    hovered: expect.objectContaining({
      type: "local-variable",
      binding: expect.objectContaining({
        name: "x",
      }),
    }),
  });
});

test("hovering with scheme", () => {
  const src = `let f = fn x, y { y }`;
  const hoverable = parseHover(src, "y", 2)!;
  if (hoverable.hovered.type !== "local-variable") {
    throw new Error("fail");
  }
  const resolved = resolveType(hoverable.hovered.binding.$type);
  if (resolved.type !== "unbound") {
    throw new Error("fail");
  }
  // expect(scheme).toEqual(
  //   expect.objectContaining({
  //     [resolved.id]: "b",
  //   }),
  // );
});

test("hover a reference to a global binding", () => {
  const src = `
        let x = 42
        let y = x
    `;
  const hoverable = parseHover(src, "x", 2)!;
  expect(hoverable).toEqual<Hovered>({
    range: rangeOf(src, "x", 2),
    hovered: expect.objectContaining({
      type: "global-variable",
      declaration: expect.objectContaining({
        binding: expect.objectContaining({
          name: "x",
        }),
      }),
    }),
  });
});

test("hover a reference to a global binding inside a fn, application, if, let", () => {
  const src = `
        let x = 42
        let y = fn {
            if true {
                let y = 0;
                f(0, x)
            } else {
                0
            }
        }
    `;
  const hoverable = parseHover(src, "x", 2)!;
  expect(hoverable).toEqual<Hovered>({
    range: rangeOf(src, "x", 2),
    hovered: expect.objectContaining({
      type: "global-variable",
      declaration: expect.objectContaining({
        binding: expect.objectContaining({
          name: "x",
        }),
      }),
    }),
  });
});

test("hover within a list", () => {
  const src = `
  let hovered_number = 42
  let expr = [hovered_number]
`;
  const hoverable = parseHover(src, "hovered_number", 2)!;
  expect(hoverable).toEqual<Hovered>({
    range: rangeOf(src, "hovered_number", 2),
    hovered: expect.objectContaining({
      type: "global-variable",
    }),
  });
});

test("hover within a pipe call", () => {
  const src = `
        let id = fn x { x }
        let hovered_number = 42
        let expr =
          hovered_number
          |> id()
          |> id()
    `;
  const hoverable = parseHover(src, "hovered_number", 2)!;
  expect(hoverable).toEqual<Hovered>({
    range: rangeOf(src, "hovered_number", 2),
    hovered: expect.objectContaining({
      type: "global-variable",
    }),
  });
});

test("hover a local binding in a match expr", () => {
  const src = `
        let m = match something {
          Constr(Nested(_, loc_var)) => 0,
        }
    `;
  const hoverable = parseHover(src, "loc_var")!;
  expect(hoverable).toEqual<Hovered>(
    expect.objectContaining({
      range: rangeOf(src, "loc_var"),
    }),
  );
});

test("hover a reference to a local binding in a match expr", () => {
  const src = `
        let m = match something {
            0 => 0,
            Constr(x) => x,
        }
    `;
  const hoverable = parseHover(src, "x", 2)!;
  expect(hoverable).toEqual<Hovered>({
    range: rangeOf(src, "x", 2),
    hovered: expect.objectContaining({
      type: "local-variable",
      binding: expect.objectContaining({
        name: "x",
      }),
    }),
  });
});

test("hover a reference to a local binding in a match expr subject", () => {
  const src = `
      let something = 42
      let m = match something { }
    `;

  const hoverable = parseHover(src, "something", 2)!;
  expect(hoverable).toEqual<Hovered>(
    expect.objectContaining({
      range: rangeOf(src, "something", 2),
    }),
  );
});

test("hover a constructor", () => {
  const src = `
        type T { Constr(Int) }
        let t = Constr(42)
    `;
  const hoverable = parseHover(src, "Constr", 2)!;
  expect(hoverable).toEqual<Hovered>({
    range: rangeOf(src, "Constr", 2),
    hovered: expect.objectContaining({
      type: "constructor",
    }),
  });
});

test("hover a type def", () => {
  const src = `
      type T { }
    `;
  const hoverable = parseHover(src, "T", 1)!;
  expect(hoverable).toEqual<Hovered>({
    range: rangeOf(src, "type T { }", 1),
    hovered: expect.objectContaining({
      type: "type",
    }),
  });
});

test("hover a type ast", () => {
  const src = `
        type T { }
        type Box<a> { }

        @extern
        @type () -> Box<T>
        let t
    `;
  const hoverable = parseHover(src, "T", 2)!;
  expect(hoverable).toEqual<Hovered>({
    range: rangeOf(src, "T", 2),
    hovered: expect.objectContaining({
      type: "type",
    }),
  });
});

test("hover a type ast in constructor", () => {
  const src = `
      type X { }
      type T { Constr(X) }
    `;
  const hoverable = parseHover(src, "X", 2)!;
  expect(hoverable).toEqual<Hovered>({
    range: rangeOf(src, "X", 2),
    hovered: expect.objectContaining({
      type: "type",
      typeDecl: expect.objectContaining({
        name: "X",
      }),
    }),
  });
});

test("hover on a field (instantiated)", () => {
  const src = `
      type X<gen> struct { some_field: gen }
      let x = X { some_field: 42 }

      pub let hov = x.some_field
    `;
  const hoverable = parseHover(src, "some_field", 3)!;

  expect(hoverable.hovered).toEqual({
    type: "field",
    type_: "Int",
  });
});

test("hover on a field (tvar)", () => {
  const src = `
      type X<gen> struct { some_field: gen }
      pub let hov = fn x { x.some_field }
    `;
  const hoverable = parseHover(src, "some_field", 2)!;

  expect(hoverable.hovered).toEqual({
    type: "field",
    type_: "a",
  });
});

test("snapshot when hovering on global fn", () => {
  const src = `let glb = fn x, y { y }`;
  const hoverable = parseHover(src, "glb")!;
  expect(hoverToMarkdown(hoverable)).toMatchSnapshot();
});

test("snapshot when hovering on an extern type", () => {
  const src = `
    @extern
    @type (x) -> x
    let glb
  `;
  const hoverable = parseHover(src, "glb")!;
  expect(hoverToMarkdown(hoverable)).toMatchSnapshot();
});

test("snapshot when hovering on an extern type reference", () => {
  const src = `
      @extern
      @type (x) -> x
      let glb

      let v = glb
    `;
  const hoverable = parseHover(src, "glb", 2)!;
  expect(hoverToMarkdown(hoverable)).toMatchSnapshot();
});

test("snapshot when hovering on a local fn", () => {
  const src = `
      let glb = fn a {
        let closure = fn x { a };
        0
      }
    `;
  const hoverable = parseHover(src, "closure")!;
  expect(hoverToMarkdown(hoverable)).toMatchSnapshot();
});

test("snapshot when hovering on a constructor", () => {
  const src = `
      type T<content> { X, Constr(content) }
    `;
  const hoverable = parseHover(src, "Constr")!;
  expect(hoverToMarkdown(hoverable)).toMatchSnapshot();
});

test("snapshot when hovering on a constructor reference", () => {
  const src = `
      type T<content> { Constr(content) }

      pub let c = Constr(42)
    `;
  const hoverable = parseHover(src, "Constr", 2)!;
  expect(hoverToMarkdown(hoverable)).toMatchSnapshot();
});

function parseHover(
  src: string,
  hovering: string,
  occurrenceNumber = 1,
): Hovered | undefined {
  const position = rangeOf(src, hovering, occurrenceNumber).start;

  if (position === undefined) {
    throw new Error("Invalid position");
  }

  const parsed = unsafeParse(src);
  const [typed, _] = typecheck("pkg", "Main", parsed);
  const ret = hoverOn("", "Ns", typed, position);
  if (ret === undefined) {
    throw new Error("Undefined hover");
  }
  return ret;
}
