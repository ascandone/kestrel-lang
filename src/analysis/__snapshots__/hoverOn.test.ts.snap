// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`snapshot when hovering on a constructor 1`] = `
"\`\`\`
Constr: (content) -> T<content>
\`\`\`
type constructor
"
`;

exports[`snapshot when hovering on a constructor reference 1`] = `
"\`\`\`
Constr: (content) -> T<content>
\`\`\`
type constructor
"
`;

exports[`snapshot when hovering on a local fn 1`] = `
"\`\`\`
closure: (a) -> b
\`\`\`
local declaration
"
`;

exports[`snapshot when hovering on an extern type 1`] = `
"\`\`\`
glb: (x) -> x
\`\`\`

"
`;

exports[`snapshot when hovering on an extern type reference 1`] = `
"\`\`\`
glb: (x) -> x
\`\`\`

"
`;

exports[`snapshot when hovering on global fn 1`] = `
"\`\`\`
glb: (a, b) -> b
\`\`\`

"
`;
